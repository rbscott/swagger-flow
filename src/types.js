// @flow

type FieldBase = {|
  name: string,

  // Unique identifier used to resolve references.
  id: string,
|};

export type SwaggerArray = {|
  items: ?(Field | Array<Field>),
  additionalItems: ?(boolean | Field),
  type: 'array',
  ...FieldBase,
|};

export type SwaggerEnum = {|
  values: Array<string>,
  type: 'enum',
  ...FieldBase,
|};

export type SwaggerObject = {|
  fields: Array<Field>,
  type: 'object',
  ...FieldBase,
|};

export type SwaggerPrimitive = {|
  // There are some refinements that could happen here such as regular expressions,
  // etc, but for now just keep it simple.
  type: (
    'boolean' |
    'integer' |
    'string'
  ),
  ...FieldBase,
|};

export type Field = (
  SwaggerEnum |
  SwaggerObject |
  SwaggerPrimitive |
  SwaggerArray
);

export type Schema = {|
  items: Array<Field>,
  lookupTable: Map<string, Field>,
|};

// Not sure if this will ever be necessary.
export type Definition = ({|
  type: 'schema',
  schema: Schema
|} | {|
  type: 'field',
  field: Field
|});

export const PRIMITIVE_TYPE_BOOLEAN = 'boolean';
export const PRIMITIVE_TYPE_INTEGER = 'integer';
export const PRIMITIVE_TYPE_NUMBER = 'number';
export const PRIMITIVE_TYPE_STRING = 'string';

export const PRIMITIVE_TYPES = [
  PRIMITIVE_TYPE_BOOLEAN,
  PRIMITIVE_TYPE_INTEGER,
  PRIMITIVE_TYPE_NUMBER,
  PRIMITIVE_TYPE_STRING,
];
