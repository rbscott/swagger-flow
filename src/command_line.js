// @flow

import { Command } from 'commander';

import type { ParserConfig } from './parser';
import type { GeneratorConfig } from './generator';

export type CommandLineRun = {|
  generatorConfig: GeneratorConfig,
  parserConfig: ParserConfig,
  swaggerPath: string,
|};

export type CommandLineOptions = CommandLineRun;

const PROGRAM_VERSION = '0.1.0';

const parseCommandLine = (argv: Array<string>): ?CommandLineOptions => {
  const program = new Command();

  program
    .version(PROGRAM_VERSION, '-v, --version')
    .usage('[options]')
    .arguments('<swagger-path>')
    .option(
      '--controller-key <controller-key>',
      'Name of the key to select the controller (e.g. x-swagger-router-controller).'
    )
    .option('--no-open-classes', 'Enable exact class definitions on types.')
    .option('--no-split-body', 'Treat a body parameter named body as just another parameter.')
    .parse(argv);

  if (program.args.length !== 1) {
    program.outputHelp();
    return null;
  }

  return {
    parserConfig: {
      controllerKey: program.controllerKey,
      splitBody: program.splitBody,
    },
    generatorConfig: {
      openClasses: program.openClasses,
    },
    swaggerPath: program.args[0],
  };
};

export default parseCommandLine;
