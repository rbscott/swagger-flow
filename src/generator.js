// @flow

import {
  PRIMITIVE_TYPE_INTEGER,
  PRIMITIVE_TYPES,
} from './types';

import type {
  Field,
  Schema,
  SwaggerArray,
  SwaggerEnum,
  SwaggerObject,
  SwaggerPrimitive,
  SwaggerReference,
} from './types';

type LookupTable = Map<string, Field>;

type Config = {|
  // Set to true to never use exact class definitions.
  openClasses: boolean,
|};

/**
 * Generate a type declaration for the given field. this is the body that would go on the
 * left side of the colon or equals sign. The caller is responsible for formatting
 * what this is assigned to.
 */
export const generateDeclaration = (
  field: Field,
  lookupTable: LookupTable,
  config: Config,
): string => {
  const { type, } = field;

  if (type === PRIMITIVE_TYPE_INTEGER) {
    return 'number';
  }

  if (PRIMITIVE_TYPES.indexOf(type) >= 0) {
    return type;
  }

  // The if statements could be a switch statement, but refinement doesn't work inside of a case
  // statement. Also, you have to use field.type, otherwise flow doesn't handle refinement
  // properly.
  if (field.type === 'allOf') {
    const { choices } = field;

    if (choices.length === 0) {
      return '/* allOf without types, using mixed. */ mixed';
    }

    // Assuming that complex types are for objects, this will not work for other types.
    const declarations = choices.map((choice) => {
      return `...${generateDeclaration(choice, lookupTable, config)}`;
    })
    .sort()
    .join(', ');

    // Not sure if exact should be used here?
    return `{ ${declarations} }`;

  // anyOf and oneOf are slightlty different, but there doesn't appear to be a way to model
  // oneOf in flow. Using the same logic for both.
  } else if (field.type === 'anyOf' || field.type === 'oneOf') {
    const { choices } = field;

    if (choices.length === 0) {
      return `/* ${field.type} without types, using mixed. */ mixed`;
    }

    // Assuming that complex types are for objects, this will not work for other types.
    const declarations = choices.map((choice) => {
      return generateDeclaration(choice, lookupTable, config);
    })
    .sort()
    .join(' | ');

    return `( ${declarations} )`;

  } else if (field.type === 'enum') {
    const values = field.values
      .sort()
      .map(value => JSON.stringify(value))
      .join(' | ');

    return `(${values})`;
  } else if (field.type === 'array') {
    const declaration = (field.items) ?
      generateDeclaration(field.items, lookupTable, config) : 'mixed';

      return `Array<${declaration}>`;

  } else if (field.type === 'ref') {
    const { ref } = field;
    const refField = lookupTable.get(ref);

    return refField ? refField.name : `/* the ref: '${ref}' is missing, defaulting to mixed */ mixed`;

  } else if (field.type === 'object') {
    // If additional properties are allowed, do not use the exact class
    // definition.
    const exact = (config.openClasses || field.additionalProperties) ? '' : '|';
    const { required } = field;

    const fields = Array.from(field.fields.entries()).map(([name, objectField]) => {
      const option = (required.indexOf(name) >= 0) ? '' : '?';
      return `${name}: ${option}${generateDeclaration(objectField, lookupTable, config)}`;
    })
    .sort()
    .join(', ');

    return `{${exact} ${fields} ${exact}}`;

  }

  return '';
};

/**
 * Given a list of items, generate a map from the ID of an Item to the field.
 */
const buildLookupTable = (items: Array<Field>): LookupTable => {
  const lookupTable = new Map();

  items.forEach((item) => {
    lookupTable.set(item.id, item);
  });

  return lookupTable;
};

const generateFlowTypes = (schema: Schema, config: Config): string => {
  const lookupTable = buildLookupTable(schema.items);

  return schema.items.reduce((accum, item) => {
    const declaration = generateDeclaration(item, lookupTable, config);

    return accum.concat(`export type ${item.name} = ${declaration};`);
  }, [])
  .join('\n\n');
};

export default generateFlowTypes;
