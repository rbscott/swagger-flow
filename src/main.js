// @flow

export type Foo = {|
  a: string
|};

const fn = (): Foo => {
  return null;
}
