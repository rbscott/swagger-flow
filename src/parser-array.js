// @flow

import type {
  Field,
  SwaggerArray,
} from './types';

// recursive import.
import { parseItemSchema } from './parser';

/**
 * Parse the schema for an Array. If the definition is not for an Array, parseArray
 * returns null.
 *
 */
const parseArray = (name: string, id: string, itemSchema: Object): ?SwaggerArray => {
  if (itemSchema.type !== 'array') {
    return null;
  }

  const {
    items,
  } = itemSchema;

  const swaggerArray = {
    id,
    items: undefined,
    name,
    type: 'array',
  };

  // For all of these cases, the name and ID will not be displayed since it isn't a root level
  // definition, but it will be useful for debugging.
  if (typeof items === 'object') {
    const itemName = `${name}_ArrayType`;
    const itemId = `${id}/ArrayType`;

    swaggerArray.items = parseItemSchema(itemName, itemId, items);
  } else {
    // Could do additional validation, technically the only valid
    // value for items is an object, and array or undefined.
    // https://swagger.io/specification/#schemaObject -> items must be an object.
  }

  return swaggerArray;
};

export default parseArray;
