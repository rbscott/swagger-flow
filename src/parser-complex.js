// @flow

import type {
  Field,
  SwaggerComplexType,
} from './types';

// recursive import.
import { parseItemSchema } from './parser';

/**
 * Parse the schema for a Complex Type. If the definition is not for a ComplexType, parseComplexType
 * returns null.
 *
 */
const parseComplexType = (name: string, id: string, itemSchema: Object): ?SwaggerComplexType => {
  const {
    allOf,
    anyOf,
    oneOf,
  } = itemSchema;

  // If none of those properties are set, then ignore it.
  let subTypes = undefined;
  let type = undefined;

  if (allOf) {
    subTypes = allOf;
    type = 'allOf';
  } else if (anyOf) {
    subTypes = anyOf;
    type = 'anyOf';
  } else if (oneOf) {
    subTypes = oneOf;
    type = 'oneOf';
  } else {
    return null;
  }

  if (!Array.isArray(subTypes)) {
    throw new Error(`${name}, ${id} Complex type is not an array ${subTypes}`);
  }

  let required = [];

  const choices = subTypes.map((subType, index) => {
    const subTypeName = `${name}_SubType_${index}`;
    const subTypeId = `${id}/SubType_${index}`;

    return parseItemSchema(subTypeName, subTypeId, subType);
  });

  if (Array.isArray(itemSchema.required)) {
    if (itemSchema.required.filter(i => typeof i !== 'string').length > 0) {
      throw new Error(`All entries in "required" must be strings ${name}, ${itemSchema.required}`);
    }

    required = itemSchema.required;
  }


  return {
    choices,
    id,
    name,
    required,
    type,
  };
};

export default parseComplexType;
