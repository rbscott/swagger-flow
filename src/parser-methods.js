// @flow

import { parseItemSchema } from './parser';
import type { ParserConfig } from './parser';

import type {
  Field,
  FieldMap,
  SwaggerObject,
} from './types';

const upperCaseFirstLetter = (value: string): string => {
  return value.replace(/^([a-z]?)/, (match, firstChar) => firstChar.toUpperCase());
}

const nameOperation = (
  prefix: string,
  path: string,
  method: string,
  operationId: ?string,
  controller: ?string,
): string => {
  if (operationId) {
    return `${prefix}${upperCaseFirstLetter(controller ? controller : '')}${upperCaseFirstLetter(operationId)}`;
  }

  return `/${path}/${method}`
    // Remove all double slashes
    .replace(/\/\/+/g, '/')
    // Change any parameters to the string For${ParameterName}
    .replace(/\{(.*?)\}/g, (match, name) => 'For' + upperCaseFirstLetter(name))
    // Remove all slashes, and then capitalize the next character.
    .replace(/\/(.)/g, (match, firstChar) => firstChar.toUpperCase())
    .replace(/^/, prefix);
}

const getParameterSchema = (parameter: Object): Object => {
  if (typeof parameter.schema === 'object' && parameter.in === 'body') {
    return parameter.schema;
  }

  return parameter;
};

const parseParametersArray = (path: string, name: string, parameters: Object): FieldMap => {
  if (!Array.isArray(parameters)) {
    return {};
  }

  return parameters
    .filter((parameter) => {
      return (typeof parameter === 'object' && parameter && parameter.type !== 'file');
    })
    .reduce((accum, parameter) => {
      if (typeof parameter.name !== 'string') {
        return accum;
      }

      const parameterName = `${name}_Field_${parameter.name}`;
      const parameterId = `#${path}/${parameter.name}`;
      const parameterSchema = getParameterSchema(parameter);

      return {
        ...accum,
        [parameter.name]: parseItemSchema(parameterName, parameterId, parameterSchema),
      }
    }, {});
}

const parseMethodParameters = (
  path: string,
  method: string,
  methodSchema: Object,
  inheritedParameters: FieldMap,
  controller: ?string,
  splitBody: boolean,
): {| main: ?SwaggerObject, body: ?Field |} => {
  const result = {
    main: undefined,
    body: undefined,
  };

  if (typeof methodSchema !== 'object') {
    return result;
  }

  if (!Array.isArray(methodSchema.parameters)) {
    return result;
  }

  const { parameters } = methodSchema;

  const [main, body] = parameters.reduce((accum, parameter) => {
    // Don't need file parameters for now.
    if (parameter.type === 'file') {
      return accum;
    }

    const [main, body] = accum;

    if (!splitBody) {
      return [main.concat(parameter), body];
    }

    if (parameter.in === 'body' && parameter.name === 'body') {
      if (body) {
        console.log('Multiple body parameters, using first one');
        return [main, body];
      }

      return [main, parameter];
    }

    return [main.concat(parameter), body];
  }, [[], undefined]);

  const name = nameOperation('Rqst', path, method, methodSchema.operationId, controller);

  const fieldsArray = main
    .map((parameter) => {
      const parameterName = `${name}_Field_${parameter.name}`;
      const parameterId = `#${path}/${method}/${parameter.name}`;
      const parameterSchema = getParameterSchema(parameter);

      return [
        parameter.name,
        parseItemSchema(parameterName, parameterId, parameterSchema)
      ];
    });

  const fields = new Map([
    ...Object.entries(inheritedParameters),
    ...fieldsArray
  ]);

  result.main = {
    // TODO: Figure out how to set this properly.
    additionalProperties: false,
    fields,
    id: `#${path}/${method}`,
    name,
    // TODO: Update this with the required fields.
    required: [],
    type: 'object',
  };

  if (body) {
    const bodySchema = getParameterSchema(body);
    const bodyName = `${name}Body`;
    const bodyId = `#${path}/${method}/body`;
    result.body = parseItemSchema(bodyName, bodyId, bodySchema);
  }

  return result;
}

const parseMethodResponses = (
  path: string,
  method: string,
  methodSchema: Object,
  controller: ?string,
): Array<Field> => {
  if (typeof methodSchema.responses !== 'object') {
    return [];
  }

  const { responses } = methodSchema;

  const choices = Object.keys(responses).reduce((accum, code) => {
    if (typeof responses[code] !== 'object' || typeof responses[code].schema !== 'object') {
      return accum;
    }

    const responseSchema  = responses[code].schema;

    if (responseSchema.type === 'file') {
      return accum;
    }

    // Append the method with the code so each response has a unique name.
    const operationId = (typeof methodSchema.operationId === 'string') ?
      `${methodSchema.operationId}${upperCaseFirstLetter(code)}` : null;

    const responseCode = nameOperation(
      'Resp',
      path,
      // Append the method with the code so each response has a unique name.
      `${method}/${code}`,
      operationId,
      controller,
    );
    const responseCodeId = `#${path}/${method}/${code}`;

    return accum.concat(parseItemSchema(responseCode, responseCodeId, responseSchema));
  }, []);

  if (choices.length === 0) {
    return [];
  }

  const responseName = nameOperation(
    'Resp',
    path,
    `${method}`,
    methodSchema.operationId,
    controller,
  );
  const responseId = `#${path}/${method}`;

  const choiceReferences = choices.map((choice) => {
    return {
      id: `${path}/choices/${choice.name}`,
      name: `${responseName}_Ref_${choice.name}`,
      type: 'ref',
      ref: choice.id,
    };
  })

  return choices.concat({
    choices: choiceReferences,
    id: responseId,
    name: responseName,
    // TODO: Figure out what should be specified here.
    required: [],
    type: 'anyOf',
  });
};

/**
 * Parse all of the parameters and responses for each request and create a schema for them.
 */
const parseMethods = (schema: Object, config: ParserConfig): Array<Field> => {
  if (typeof schema.paths !== 'object') {
    return [];
  }

  const { paths } = schema;
  const {
    controllerKey,
    splitBody,
  } = config;

  const globalParameters = parseParametersArray('/parameters', 'parameters', schema.parameters);

  return Object.keys(paths).reduce((accum1, path) => {
    if (typeof paths[path] !== 'object') {
      return accum1;
    }

    const definition = paths[path];
    const pathController = (typeof definition[controllerKey] === 'string') ?
      definition[controllerKey] : undefined;

    const pathName = nameOperation('', path, '', undefined);

    const inheritedParameters = {
      ...globalParameters,
      ...parseParametersArray(path, pathName, definition.parameters),
    };

    return Object.keys(definition).reduce((accum2, method) => {
      const methodSchema = definition[method];

      if (typeof methodSchema !== 'object' || typeof method !== 'string') {
        return accum1;
      }

      const methodController = (typeof methodSchema[controllerKey] === 'string') ?
        methodSchema[controllerKey] : pathController;

      const {
        main,
        body,
      } = parseMethodParameters(
        path, method, methodSchema, inheritedParameters, methodController, splitBody,
      );
      const methodResponses = parseMethodResponses(path, method, methodSchema, methodController);
      return accum2.concat(main || [], body || [], methodResponses);
    }, accum1);

  }, []);
};

export default parseMethods;
