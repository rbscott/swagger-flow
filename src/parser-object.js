// @flow

import type {
  Field,
  SwaggerObject,
} from './types';

// recursive import.
import { parseItemSchema } from './parser';

/**
 * Parse the schema for a primitive type. If the definition is not for a primitive type,
 * return null.
 */
const parseObject = (name: string, id: string, itemSchema: Object): ?SwaggerObject => {
  // Treat anything that has properties set (or a type object) as an object.
  if (itemSchema.type !== 'object' && typeof itemSchema.properties !== 'object') {
    return null;
  }

  const {
    additionalProperties,
    properties,
    required,
  } = itemSchema;

  const fields = new Map();

  Object.keys(properties || {}).forEach((key) => {
    const fieldSchema = properties[key];

    const fieldName = `${name}_Field_${key}`;
    const fieldId = `${id}/Field_${key}`;

    fields.set(key, parseItemSchema(fieldName, fieldId, fieldSchema));
  });

  const swaggerObject = {
    additionalProperties: undefined,
    fields,
    id,
    name,
    required: [],
    type: 'object',
  };

  if (typeof additionalProperties === 'boolean') {
    swaggerObject.additionalProperties = additionalProperties;
  } else if (typeof additionalProperties === 'object') {
    const addName = `${name}_AdditionalPropType`;
    const addId = `${id}/AdditionalPropType`;

    swaggerObject.additionalProperties = parseItemSchema(addName, addId, additionalProperties);
  }

  if (Array.isArray(required)) {
    if (required.filter(i => typeof i !== 'string').length > 0) {
      throw new Error(`All entries in "required" must be strings ${name}, ${required}`);
    }

    swaggerObject.required = required;
  }

  return swaggerObject;
};

export default parseObject;
