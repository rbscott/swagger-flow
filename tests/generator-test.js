
import { generateDeclaration } from '../src/generator';
import { PRIMITIVE_TYPES } from '../src/types';

describe('generateDeclaration', () => {
  const DEFAULT_CONFIG = { openClasses: true };
  const STRICT_CONFIG = { openClasses: false };

  /*******
   * array
   *******/

  describe('array', () => {
    it(`Should build an array declaration.`, () => {
      const field = {
        id: '#definitions/Array',
        items: {
          id: 'unused',
          name: 'unused',
          type: 'boolean',
        },
        name: 'Array',
        type: 'array',
      };

      expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
    });

    it(`Should use mixed if items is undefined.`, () => {
      const field = {
        id: '#definitions/Array',
        items: undefined,
        name: 'Array',
        type: 'array',
      };

      expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
    });

    it(`Should build an array declaration with a reference type.`, () => {
      const lookupTable = new Map(Object.entries({
        '#definitions/Count': {
          id: '#definitions/Count',
          name: 'Count',
          type: 'integer'
        },
      }));

      const field = {
        id: '#definitions/Array',
        items: {
          id: 'unused',
          name: 'unused',
          ref: '#definitions/Count',
          type: 'ref',
        },
        name: 'Array',
        type: 'array',
      };

      expect(generateDeclaration(field, lookupTable, DEFAULT_CONFIG)).toMatchSnapshot();
    });

  });

  /*******
   * enum
   *******/

  describe('enum', () => {
    it(`Should build a basic declaration for an enum.`, () => {
      const field = {
        id: '#definitions/Enum',
        name: 'Enum',
        type: 'enum',
        values: ['value1', 'value2', 'value3'],
      };

      expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
    });

    it(`Should sort the enum values.`, () => {
      const field = {
        id: '#definitions/Enum',
        name: 'Enum',
        type: 'enum',
        values: ['red', 'green', 'blue'],
      };

      expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
    });
  });

  /*******
   * object
   *******/

  describe('object', () => {
    it(`Should build a basic object declaration.`, () => {
      const fields = new Map(Object.entries({
        field1: {
          id: 'unused',
          name: 'unused',
          type: 'boolean',
        },
      }));

      const field = {
        additionalProperties: false,
        fields: fields,
        id: '#definitions/Object',
        name: 'Object',
        required: [],
        type: 'object',
      };

      expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
    });

    it(`Should build an exact object declaration.`, () => {
      const fields = new Map(Object.entries({
        field1: {
          id: 'unused',
          name: 'unused',
          type: 'boolean',
        },
      }));

      const field = {
        additionalProperties: false,
        fields: fields,
        id: '#definitions/Object',
        name: 'Object',
        required: [],
        type: 'object',
      };

      expect(generateDeclaration(field, new Map(), STRICT_CONFIG)).toMatchSnapshot();
    });

    it(`Should build a loose object declaration if additionalProperties is set.`, () => {
      const fields = new Map(Object.entries({
        field1: {
          id: 'unused',
          name: 'unused',
          type: 'boolean',
        },
      }));

      const field = {
        additionalProperties: true,
        fields: fields,
        id: '#definitions/Object',
        name: 'Object',
        required: [],
        type: 'object',
      };

      expect(generateDeclaration(field, new Map(), STRICT_CONFIG)).toMatchSnapshot();
    });

    it(`Should build a basic object with multiple fields declaration.`, () => {
      const fields = new Map(Object.entries({
        field1: {
          id: 'unused',
          name: 'unused',
          type: 'boolean',
        },
        field2: {
          id: 'unused',
          name: 'unused',
          type: 'boolean',
        },
      }));

      const field = {
        additionalProperties: false,
        fields: fields,
        id: '#definitions/Object',
        name: 'Object',
        required: [],
        type: 'object',
      };

      expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
    });

    it(`Should build a nested object with multiple fields declaration.`, () => {
      const fields = new Map(Object.entries({
        field1: {
          id: 'unused',
          name: 'unused',
          type: 'boolean',
        },
        field2: {
          additionalProperties: false,
          fields: new Map(Object.entries({
            nestedField1: {
              id: 'unused',
              name: 'unused',
              type: 'string',
            },
          })),
          id: '#definitions/Object',
          name: 'Object',
          required: [],
          type: 'object',
        },
      }));

      const field = {
        additionalProperties: false,
        fields: fields,
        id: '#definitions/Object',
        name: 'Object',
        required: [],
        type: 'object',
      };

      expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
    });

  });

  /*******
   * primitives
   *******/

  describe('primitives', () => {
    PRIMITIVE_TYPES.forEach((type) => {
      it(`Should build a declaration for a ${type} primitive`, () => {
        const field = {
          id: '#definitions/Primitive',
          name: 'Primitive',
          type,
        };

        expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
      });
    });
  });

  /*******
   * ref
   *******/

  describe('ref', () => {
    it(`Should build a declaration for a ref.`, () => {
      const lookupTable = new Map(Object.entries({
        '#definitions/Count': {
          id: '#definitions/Count',
          name: 'Count',
          type: 'integer'
        },
      }));

      const field = {
        id: '#definitions/Ref',
        name: 'Ref',
        ref: '#definitions/Count',
        type: 'ref',
      };

      expect(generateDeclaration(field, lookupTable, DEFAULT_CONFIG)).toMatchSnapshot();
    });

    it(`Should use "mixed" if the ref doesn't exist.`, () => {
      const field = {
        id: '#definitions/Ref',
        name: 'Ref',
        ref: '#definitions/Missing',
        type: 'ref',
      };

      expect(generateDeclaration(field, new Map(), DEFAULT_CONFIG)).toMatchSnapshot();
    });

  });
});
