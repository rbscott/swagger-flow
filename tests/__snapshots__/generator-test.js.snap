// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDeclaration array Should build an array declaration with a reference type. 1`] = `"Array<Count>"`;

exports[`generateDeclaration array Should build an array declaration. 1`] = `"Array<boolean>"`;

exports[`generateDeclaration array Should use mixed if items is undefined. 1`] = `"Array<mixed>"`;

exports[`generateDeclaration enum Should build a basic declaration for an enum. 1`] = `"(\\"value1\\" | \\"value2\\" | \\"value3\\")"`;

exports[`generateDeclaration enum Should sort the enum values. 1`] = `"(\\"blue\\" | \\"green\\" | \\"red\\")"`;

exports[`generateDeclaration object Should build a basic object declaration. 1`] = `"{ field1: ?boolean }"`;

exports[`generateDeclaration object Should build a basic object with multiple fields declaration. 1`] = `"{ field1: ?boolean, field2: ?boolean }"`;

exports[`generateDeclaration object Should build a loose object declaration if additionalProperties is set. 1`] = `"{ field1: ?boolean }"`;

exports[`generateDeclaration object Should build a nested object with multiple fields declaration. 1`] = `"{ field1: ?boolean, field2: ?{ nestedField1: ?string } }"`;

exports[`generateDeclaration object Should build an exact object declaration. 1`] = `"{| field1: ?boolean |}"`;

exports[`generateDeclaration primitives Should build a declaration for a boolean primitive 1`] = `"boolean"`;

exports[`generateDeclaration primitives Should build a declaration for a integer primitive 1`] = `"number"`;

exports[`generateDeclaration primitives Should build a declaration for a number primitive 1`] = `"number"`;

exports[`generateDeclaration primitives Should build a declaration for a string primitive 1`] = `"string"`;

exports[`generateDeclaration ref Should build a declaration for a ref. 1`] = `"Count"`;

exports[`generateDeclaration ref Should use "mixed" if the ref doesn't exist. 1`] = `"/* the ref: '#/definitions/Missing' is missing, defaulting to mixed */ mixed"`;
